### PR TITLE
CRAYSAT-1530: Document `sat bootprep` improvements

### PR DESCRIPTION
**Pull request by haasken-hpe**
## Summary and Scope

Document the `sat bootprep` improvements in SAT 2.4. This includes:

* Variable substitution * HPC CSM Sofware Recipe variables (product versions) * Dynamic variable for base image name in image names * Dynamic variable for image name in session template names
* New `ref_name` and `image_ref` properties of images.
* New schema for specifying base in image
* New schema for specifying image in session template
* Using recipes and images provided by products
* New `schema_version` property of `sat bootprep` input files.

Also cleaned up examples a little by removing some example file contents that show where sections belong in the file. They are not necessary since the order of the sections doesn't really matter, and we provide a complete example near the end of the usage already, which shows how the sections appear together in the input file. This also allows us to use YAML syntax highlighting in more example input file snippets.

## Issues and Related PRs

* Resolves [CRAYSAT-1530](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1530)

### Tested on:

  * GitHub

### Test description:
Examined in GitHub to ensure everything rendered correctly and links work.

## Risks and Mitigations

None

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -35,15 +35,39 @@ three main sections, one each for configurations, images, and session templates.
...
-configurations:
+schema_version: 1.0.2
 ```
 
+To get the current `sat bootprep` input file schema version understood by the
+version of `sat` installed on the system, look at the value of the `version` key
+in the schema definition:
+
+```screen
+ncn-m001# sat bootprep view-schema | grep -E '^version:'
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:54]: Yeah, I guess not.

File: docs/usage.md [hunk 2]
```diff
@@ -35,15 +35,39 @@ three main sections, one each for configurations, images, and session templates.
...
+version: '1.0.2'
+```
+
+The `sat bootprep run` command will validate the schema version specified in the
+input file and make sure it is compatible with the schema version understood by
+the current version of `sat bootprep`. See the description of the
+`schema_version` property in the JSON Schema file for more details about schema
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:61]: Yeah, I'll just drop JSON Schema here. I think I'll say, "See the description of the `schema_version` property in the bootprep input file schema for more details about schema version validation."

Then I'll refer to the sections "Viewing the exact schema specification" and "Generating user-friendly documentation".

File: docs/usage.md [hunk 3]
```diff
@@ -77,7 +101,7 @@ intended version of the Ansible playbooks. A commit hash may be specified by rep
 
 ```screen
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:102]: Yeah, the ellipses get highlighted weirdly in that case, IIRC from Jack's original PR for these instructions. Having said that, it might be more helpful to just show an entire CFS configuration here so it's easier to see that this is a layer? I think I'll do that.

[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:102]: It was a little more complicated than that because the following paragraph assumes that these two layers were in a single configuration. I ended up combining these two snippets into one YAML snippet.

File: docs/usage.md [hunk 4]
```diff
@@ -111,93 +135,115 @@ above might look something like the following:
...
+substitutions](#dynamic-variable-substitutions).
+
+Images may also contain a `description` key. It is not currently used, but it
+may be useful to describe the image in the bootprep input file.
+
+Here is an example of an image using an existing IMS recipe as its base. This
+example will build an IMS image from that recipe. It will then configure it with
+a CFS configuration named `example-compute-config`. The `example-compute-config`
+CFS configuration can be defined under the `configurations` key in the same
+input file, or it can be a CFS configuration that already exists in CFS. Running
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:177]: Yeah, it's a bit of a mouthful. I'll say "an existing CFS configuration" instead.

File: docs/usage.md [hunk 5]
```diff
@@ -111,93 +135,115 @@ above might look something like the following:
...
+  configuration: example-compute-config
+  configuration_group_names:
+  - Compute
 ```
 
-Images may also contain a `configuration` property in their definition, which
-specifies a configuration with which to customize the built image prior to
-booting. If a configuration is specified, then configuration groups must also
-be specified using the `configuration_group_names` property.
+Here is an example showing the definition of two images. The first image is
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:195]: Maybe I'm missing something, but I'm actually not showing the dynamic variable substitution here. I did that intentionally to try to gradually introduce topics.

File: docs/usage.md [hunk 6]
```diff
@@ -207,105 +253,327 @@ which are optional:
...
+Jinja2 templating of the `sat bootprep` input file is performed individually
+for each supported property. That is, the entire bootprep input file is not
+rendered by the Jinja2 template engine.
+
+The following values support rendering as a Jinja2 template:
+
+- The `name` of each configuration under the `configurations` key.
+- The values of the following keys of each layer under the `layers` key in a
+  configuration:
+    - `name`
```
[the remark this answers is not in the document]
Comment (ekamin-hpe) [reply docs/usage.md:334]: @haasken-hpe and I talked offline about this and decided we'll allow the MD030 and MD007 violations for now, while I continue to work on adding the lint configuration to docs-sat. In both of these cases we think there's a good case for configuring the lint rules to prefer these styles.

[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:334]: I discussed this with Eli a bit in slack. Both MD007 and MD030 are configurable. I don't feel too strongly about 2 spaces vs. 4 spaces for MD007. MD030 was an intentional choice that we made when working with Ben. We wanted a consistent visual indent for all multi-line list items, and we felt that 3 spaces was a little annoying to maintain in an editor, e.g. when the tab stop was set to 2 or 4 spaces. We'll discuss those two items in scrum and decide as part of Eli's CRAYSAT-1571 work.

I will address MD012 here.

**Comment by haasken-hpe** [comment]
I addressed all of Eli's comments except for the potential markdownlint violations and the comment about dynamic variable substitutions, which I didn't quite follow in commit c27614d1b9036db6048dc69e58f9389a9a0ae09e.

File: docs/usage.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ekamin-hpe) [reply docs/usage.md:197]: Nope, that was my fault. I was confusing the image references with dynamic variable substitution.

**Review by jack-stanek-hpe — APPROVED** [review]

File: docs/usage.md [hunk 7]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
-specified directly through BOS boot sets. More information can be found in the
-[CSM documentation on session
-templates](https://cray-hpe.github.io/docs-csm/en-10/operations/boot_orchestration/session_templates/).
-Additional properties not listed are passed through to the BOS session template
-as written.
-  
-An example session template might look like the following:
+As mentioned above, the parameters under `bos_parameters` are passed through
+directly to BOS. Refer to the section **BOS Session Templates** of the [*Cray
+System Management Documentation*](https://cray-hpe.github.io/docs-csm/) for more
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:262]: This was actually intentionally not a link to the specific section because we don't want to have to maintain the `en-13` part of the link. There is some discussion about this (in the other direction, from docs-csm to docs-sat) here: https://github.com/Cray-HPE/docs-csm/pull/2366#discussion_r966317466

Jessica says it's the approach techpubs is taking for now for linking between different products' documentation.

**Comment by haasken-hpe** [comment]
Once @jld-dehn-hpe completes her review, I'm going to rebase this on the latest integration to get the link checker, markdown lint, and spellcheck workflows to run. I'll also resolve the conflicts in docs/usage.md at that time.

File: docs/usage.md [hunk 8]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
+  under the `session_templates` key. The value of this variable is the name of
+  the IMS image used in this session template.
+
+These variables reduce the need to duplicate values throughout the `sat
+bootprep` input file and also make the following use cases possible:
+
+- Using the name of the recipe in the name of an image built from a recipe
+  provided by a product.
+- Using the name of the image in the name of a session template that references an
+  image built from a recipe provided by a product.
```
Comment (jld-dehn-hpe) [inline docs/usage.md:491]:
```suggestion
- You use the name of the recipe within the name of an image, and that image is
built from a recipe provided by a product.
- You use the name of the image within the name of a session template, and that
session template references an image that is built from a recipe provided by a product.
```

File: docs/usage.md [hunk 9]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
+
+### Example bootprep input files
+
+This section provides an example bootprep input file and instructions for
+obtaining the default bootprep input files delivered with a release of the HPC
+CSM Software Recipe.
```
Comment (jld-dehn-hpe) [inline docs/usage.md:497]:
```suggestion
This section provides an example bootprep input file. It also gives
instructions for obtaining the default bootprep input files delivered
with a release of the HPC CSM Software Recipe.
```

File: docs/usage.md [hunk 10]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
         node_roles_groups:
         - Compute
+        rootfs_provider_passthrough: "dvs:api-gw-service-nmn.local:300:hsn0,nmn0:0"
 ```
 
-#### Creating a pre-populated example bootprep input file
+#### Accessing default bootprep input files
+
+Default bootprep input files are delivered by the HPC CSM Software Recipe
+product. These files can be accessed by cloning the `hpc-csm-software-recipe`
```
Comment (jld-dehn-hpe) [inline docs/usage.md:557]:
```suggestion
product. You can access these files by cloning the `hpc-csm-software-recipe`
```

File: docs/usage.md [hunk 11]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
+repository.
 
-It is possible to create an example bootprep input file using values from the
-system's product catalog using the `sat bootprep generate-example` command.
+Follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
+variables](#viewing-hpc-csm-software-recipe-variables).
+
+Access the files in the `bootprep` directory of that repository:
```
Comment (jld-dehn-hpe) [inline docs/usage.md:563]: I would consider duplicating steps 1-7 here and adding accessing the `bootprep` directory as step 8. For new users, this file can largely be read from top to bottom. However, we'll break that flow if we don't include all the steps under this heading. With a large part of the information in this file new for Papaya, I think we want to avoid that.

File: docs/usage.md [hunk 12]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
 ```
 
-This file should be reviewed and edited to match the desired parameters of the
-configurations, images, and session templates.
+#### Generating an example bootprep input file
+
+Note: the `sat bootprep generate-example` command has not been updated for
+recent bootprep schema changes, and it is recommended to use the default
+bootprep input files described in [Accessing default bootprep input
+files](#accessing-default-bootprep-input-files).
```
Comment (jld-dehn-hpe) [inline docs/usage.md:574]:
```suggestion
**Note:** The `sat bootprep generate-example` command was not updated for
recent bootprep schema changes. It is recommended that you instead use the
default bootprep input files described in [Accessing default bootprep input
files](#accessing-default-bootprep-input-files).
```

File: docs/usage.md [hunk 13]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
+
```
Comment (jld-dehn-hpe) [inline docs/usage.md:575]: I would suggest removing line 575 so all the information in this section is one paragraph.

File: docs/usage.md [hunk 14]
```diff
@@ -341,15 +611,15 @@ properties:
 The raw schema definition can be difficult to understand without experience
 working with JSON Schema specifications. For this reason, a feature was included
 which can generate user-friendly HTML documentation for the input file schema
-which can be browsed with the user's preferred web browser. 
+which can be browsed with the user's preferred web browser.
```
Comment (jld-dehn-hpe) [inline docs/usage.md:614]: I would reword lines 612 through 614 as the following:
```suggestion
working with JSON Schema specifications. For this reason, a feature is included
that generates user-friendly HTML documentation for the input file schema
and can be browsed with your preferred web browser.
```

**Review by jld-dehn-hpe — CHANGES_REQUESTED** [review]

File: docs/usage.md [hunk 15]
```diff
@@ -35,15 +35,43 @@ three main sections, one each for configurations, images, and session templates.
 These sections may be specified in any order, and any of the sections may be
 omitted if desired.
 
-#### Creating CFS configurations
+#### Bootprep input file schema version
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:38]: Yeah, I was considering that as I also didn't like the unnecessary 5 levels of headings. I'll make that change.

[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:38]: Do you think it could even be shortened to "Providing a schema version"? It's a subsection of "Editing a bootprep input file" already, so I think the context is fairly clear.

File: docs/usage.md [hunk 16]
```diff
@@ -35,15 +35,43 @@ three main sections, one each for configurations, images, and session templates.
...
 
-```screen
+```yaml
 ---
-configurations:
+schema_version: 1.0.2
 ```
 
+To get the current `sat bootprep` input file schema version understood by the
+version of `sat` installed on the system, look at the value of the `version` key
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:50]: This suggestion feels less accurate to what I'm trying to say. What if we just simplify it to:

```suggestion
The current `sat bootprep` input file schema version can be viewed with the
following command:
```

I'm really just trying to convey how to get the version of the schema that is currently understood by the version of `sat bootprep` installed on the system, so that the user can use this schema version in their input files. I'm also working on providing the schema version in the bootprep input files generated by `sat bootprep generate-example` in [CRAYSAT-1545](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1545), which should make this easier for the end user.

File: docs/usage.md [hunk 17]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
 
-Putting together all of the previous input file sections, an example bootprep input
-file might look something like the following.
+#### HPC CSM Software Recipe variable substitutions
+
+The HPC CSM Software Recipe provides a manifest defining the versions of each
+HPC software product included in the recipe. These product versions can be used
+in the `sat bootprep` input file using Jinja2 templating syntax.
+
+##### Selecting the HPC CSM Software Recipe version
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:311]: Shouldn't it be "an HPC CSM Software Recipe version" due to the pronunciation of the HPC initialism?

File: docs/usage.md (hunk 8 above)
Comment (haasken-hpe) [reply docs/usage.md:491]: I think the suggestion breaks the intended meaning a little bit. I'll think about how to reword.

File: docs/usage.md (hunk 11 above)
Comment (haasken-hpe) [reply docs/usage.md:563]: I really want to avoid duplicating info and having to maintain it in two places.

There's also duplicate information in the Getting Started Guide for this, and I think that will be the more visible location for this information.

Eventually, I would like to add a `sat bootprep` subcommand that outputs the variables available from the HPC CSM Software Recipe, so that other procedure would not be necessary either. See [CRAYSAT-1532](https://jira-pro.its.hpecorp.net:8443/browse/CRAYSAT-1532) for that.

File: docs/usage.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (jld-dehn-hpe) [reply docs/usage.md:38]: Yes, I think that's a great idea!

File: docs/usage.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (jld-dehn-hpe) [reply docs/usage.md:50]: I like this change, but I think you'll also want to delete line 51, then.

File: docs/usage.md (hunk 11 above)
Comment (jld-dehn-hpe) [inline docs/usage.md:563]:
```suggestion
To do this, follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
variables](#viewing-hpc-csm-software-recipe-variables). Then, access the files in the
`bootprep` directory of that repository:
```

Comment (jld-dehn-hpe) [reply docs/usage.md:563]: I'm fine not duplicating the instructions until we have an idea of the repo setup with unified docs. However, I'd suggest changing the wording a little (see next comment).

File: docs/usage.md [hunk 18]
```diff
@@ -207,105 +255,326 @@ which are optional:
...
+  layers:
+  - name: cos-compute-integration-{{cos.version}}
+    playbook: cos-compute.yaml
+    product:
+      name: cos
+      version: "{{cos.version}}"
+      branch: integration-{{cos.version}}
+```
+
+Note that when the value of a key in the bootprep input file is a Jinja2
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:450]: The comma seems off. How about I make it a bold **Note:** like another suggestion you had?

File: docs/usage.md [hunk 19]
```diff
@@ -58,31 +86,30 @@ cray-product-catalog Kubernetes ConfigMap as necessary. A version may be
 supplied, but if it is absent, the version is assumed to be the latest version
 found in the cray-product-catalog.
 
-```screen
+Alternatively, a configuration layer may be specified by explicitly referencing
+the desired configuration repository, along with the branch containing the
+intended version of the Ansible playbooks. A commit hash may be specified by
+replacing `branch` with `commit`.
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/usage.md:92]: Applied this one manually due to required changes in surrounding context and will push a commit momentarily.

File: docs/usage.md (hunk 14 above)
Comment (haasken-hpe) [reply docs/usage.md:616]: the re-wording says "a feature ... can be browsed with your preferred web browser" which is confusing IMO. Perhaps splitting into two sentences would address your concern as well. E.g:

```
The raw schema definition can be difficult to understand without experience
working with JSON Schema specifications. For this reason, a feature is included
that generates user-friendly HTML documentation for the input file schema. This
HTML documentation can be browsed with your preferred web browser.
```

File: docs/sat_bootprep.md [hunk 20]
```diff
@@ -0,0 +1,643 @@
...
+[Generating user-friendly documentation](#generating-user-friendly-documentation).
+
+The default `sat bootprep` input files provided by the `hpc-csm-software-recipe`
+release distribution already contain the correct schema version.
+
+### Defining CFS configurations
+
+The CFS configurations are defined under a `configurations` key. Under this
+key, the user can list one or more configurations to create. For each
+configuration, a name should be given in addition to the list of layers which
```
Comment (jld-dehn-hpe) [inline docs/sat_bootprep.md:72]:
```suggestion
key, you can list one or more configurations to create. For each
configuration, give a name in addition to the list of layers that
```

File: docs/README.md [hunk 21]
```diff
@@ -36,9 +36,9 @@
 - [SAT Kibana Dashboards](dashboards/SAT_Kibana_Dashboards.md)
 - [SAT Grafana Dashboards](dashboards/SAT_Grafana_Dashboards.md)
 
-## [SAT Usage](usage.md)
+## SAT Usage
 
-- [SAT Bootprep](usage.md#sat-bootprep)
+- [SAT Bootprep](sat_bootprep.md)
```
Comment (jld-dehn-hpe) [inline docs/README.md:41]: I'd suggest removing lines 39 and 40 and having line 41 be its own level 2 heading. It doesn't really make sense to have a single topic under "SAT Usage" right now. However, I'd be all for adding the heading back if we have more topics dealing with usage besides SAT bootprep in the future.

**Review by jld-dehn-hpe — CHANGES_REQUESTED** [review]

File: docs/README.md (hunk 21 above)
Comment (haasken-hpe) [reply docs/README.md:41]: What are you envisioning the content would be under this level-2 heading in that case? I think a level-2 heading with no content would look a bit odd, so are you proposing that I add links into the specific sections of that document? Something like this?

```
- [SAT bootprep vs SAT bootsys](sat_bootprep.md#sat-bootprep-vs-sat-bootsys)
- [Editing a bootprep input file](sat_bootprep.md#editing-a-bootprep-input-file)
- [Example bootprep input files](sat_bootprep.md#example-bootprep-input-files)
- [Viewing built-in generated documentation](sat_bootprep.md#viewing-built-in-generated-documentation)
```

It seems like it's maybe exposing too much detail at this top-level of the SAT docs. Thoughts?

File: docs/sat_bootprep.md [hunk 22]
```diff
@@ -0,0 +1,643 @@
...
+
+The `sat bootprep run` command validates the schema version specified
+in the input file. The command also makes sure that the schema version
+of the input file is compatible with the schema version understood by the
+current version of `sat bootprep`. For more information on schema version
+validation, see the `schema_version` property description in the bootprep
+input file schema. For more information on viewing the bootprep input file
+schema in either raw form or user-friendly HTML form, see [Viewing the exact
+schema specification](#viewing-the-exact-schema-specification) or
+[Generating user-friendly documentation](#generating-user-friendly-documentation).
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:63]: This one renders okay. As discussed in scrum, I think it's okay to have the hard line breaks (newline characters) in the link text, but it is not okay to introduce extra whitespace (space, tabs, newlines) in between the closing square bracket `]` and the opening parenthesis `(`. It's also not okay to have whitespace in the link destination.

Here's a reference to the GitHub-flavored markdown spec: https://github.github.com/gfm/#links

File: docs/sat_bootprep.md [hunk 23]
```diff
@@ -0,0 +1,643 @@
...
+Images may also contain the following keys:
+
+- Use a `configuration` key to specify a CFS configuration with which to
+customize the built image. If a configuration is specified, then configuration
+groups must also be specified using the `configuration_group_names` key.
+- Use a `ref_name` key to specify a unique name that can refer to this image
+within the input file in other images or in session templates. The `ref_name`
+key allows references to images from the input file that have dynamically
+generated names as described in [Dynamic variable substitutions]
+(#dynamic-variable-substitutions).
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:163]: Good catch, this one was broken.

[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:163]: I'm also noticing now that I should fix up some of the spacing on these multi-line list items. I think our style is to have the start of the text line up on each subsequent line of a list item, so some of these lines should be indented by 2 spaces. I'll make that change as well.

File: docs/sat_bootprep.md [hunk 24]
```diff
@@ -0,0 +1,643 @@
...
+`boot_sets`. Each boot set can contain the following keys, all of
+which are optional:
+
+- `kernel_parameters`: the parameters passed to the kernel on the command line
+- `network`: the network over which the nodes will boot
+- `node_list`: nodes to add to the boot set
+- `node_roles_groups`: HSM roles to add to the boot set
+- `node_groups`: HSM groups to add to the boot set
+- `rootfs_provider`: the root file system provider
+- `rootfs_provider_passthrough`: parameters to add to the `rootfs=` kernel
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:244]: Huh, I wonder what happened here. Might have been something to do with applying suggestions. Good catch, I'll fix this up.

File: docs/sat_bootprep.md [hunk 25]
```diff
@@ -0,0 +1,643 @@
...
+- Use a `node_roles_groups` key to specify the HSM roles to add to the boot set.
+- Use a `node_groups` key to specify the HSM groups to add to the boot set.
+- Use a `rootfs_provider` key to specify the root file system provider.
+- Use a `rootfs_provider_passthrough` key to specify the parameters to add to the `rootfs=`
+kernel parameter.
+
+As mentioned above, the parameters under `bos_parameters` are passed through
+directly to BOS. For more information on the properties of a BOS boot set,
+refer to the section **BOS Session Templates** in the [*Cray
+System Management Documentation*](https://cray-hpe.github.io/docs-csm/).
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:257]: This one's okay as well since the newline occurs in the link title.

File: docs/sat_bootprep.md [hunk 26]
```diff
@@ -0,0 +1,643 @@
...
+
+### Dynamic variable substitutions
+
+In addition to the product version variables provided by the HPC CSM Software
+Recipe as described in [HPC CSM Software Recipe variable
+substitutions](#hpc-csm-software-recipe-variable-substitutions), additional
+variables are available. These variables are dynamic in that their values are
+determined at run-time based on the context in which they appear. The additional
+variables are as follows:
+
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:470]: Seems like another suggestion application error. Good catch.

File: docs/sat_bootprep.md [hunk 27]
```diff
@@ -0,0 +1,643 @@
...
+Additional variables are available besides the product version variables
+provided by the HPC CSM Software Recipe. (See [HPC CSM Software Recipe variable
+substitutions](#hpc-csm-software-recipe-variable-substitutions) for more
+information.) These additional variables are dynamic because their values are
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:474]: This one's okay as-is also

File: docs/sat_bootprep.md [hunk 28]
```diff
@@ -0,0 +1,643 @@
...
+
+### Accessing default bootprep input files
+
+Default bootprep input files are delivered by the HPC CSM Software Recipe
+product. You can access these files by cloning the `hpc-csm-software-recipe`
+repository.
+
+To do this, follow steps 1-7 of the procedure in [Viewing HPC CSM Software Recipe
+variables](#viewing-hpc-csm-software-recipe-variables). Then, access the files in the
+`bootprep` directory of that repository:
```
[the remark this answers is not in the document]
Comment (haasken-hpe) [reply docs/sat_bootprep.md:562]: This one's okay as-is also.

File: docs/README.md (hunk 21 above)
Comment (haasken-hpe) [reply docs/README.md]: I pushed the change described in my comment above. Let me know what you think.

Comment (jld-dehn-hpe) [reply docs/README.md]: These changes are exposing a level of detail, but I think that might actually be a good thing. My understanding is that SAT bootprep is pretty important to the install and upgrade process, and there were quite a few changes we made to this information for Papaya. This might help users understand and better navigate what's in this section when they first come to the docs.

File: docs/README.md [hunk 29]
```diff
@@ -36,9 +36,12 @@
...
 - [SAT Grafana Dashboards](dashboards/SAT_Grafana_Dashboards.md)
 
-## [SAT Usage](usage.md)
+## [SAT Bootprep](sat_bootprep.md)
 
-- [SAT Bootprep](usage.md#sat-bootprep)
+- [SAT bootprep vs SAT bootsys](sat_bootprep.md#sat-bootprep-vs-sat-bootsys)
+- [Editing a bootprep input file](sat_bootprep.md#editing-a-bootprep-input-file)
+- [Example bootprep input files](sat_bootprep.md#example-bootprep-input-files)
+- [Viewing built-in generated documentation](sat_bootprep.md#viewing-built-in-generated-documentation)
```
Comment (jld-dehn-hpe) [inline docs/README.md:44]: I'm noticing that we generally use title-style capitalization for our section headings. However, there are a few places that we use sentence-style capitalization instead. For example, [Remove obsolete configuration file sections](install.md/#remove-obsolete-configuration-file-sections) and the headings within [SAT Release Notes](release_notes.md).

We should try to be consistent, so you might want to change your headings here and in the SAT bootprep file to title-style capitalization instead. I can try to update the other headings to title-style capitalization in the future.

**Review by jld-dehn-hpe — APPROVED** [review]

File: docs/README.md (hunk 21 above)
Comment (haasken-hpe) [reply docs/README.md]: Sounds good. I'll leave it.

File: docs/README.md (hunk 29 above)
Comment (haasken-hpe) [reply docs/README.md]: Sure, fixed with 1770352b652ea3d13b83cb85e755d088da1e666c
